### PR TITLE
Enable Direct Download for System VMs

### PR DESCRIPTION
**Pull request by shwstppr**
## Description
As of now provisioning of System VMs with direct download enabled template was not supported. Even after setting such templates as SYSTEM template, while provisioning SSVM and CPVM the server would return an error message like,

> System vm template is not ready at data center 1, wait until it is ready to launch secondary storage vm

This behaviour has been changed and it allows the server to provision SSVM and CPVM even with direct download system templates.
Additionally, changes have been made to make different timeouts for direct download functionality configurable using global settings. Following global configurations have been added for this,
`direct.download.connect.timeout` - Connection establishment timeout in milliseconds for direct download
`direct.download.socket.timeout` - Socket timeout (SO_TIMEOUT) in milliseconds for direct download
`direct.download.connection.request.timeout` - Requesting a connection from connection manager timeout in milliseconds for direct download

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):
New global settings,
Screenshot from 2020-02-13 14-21-14

## How Has This Been Tested?

Admin registers a direct download private USER or ROUTING type template,

```
> register template url=http://staging.yadav.xyz/systemvmtemplate/4.14/systemvmtemplate-4.14.0-kvm.qcow2.bz2 name=NewTemplate displaytext=NewTemplate zoneids=94ecae8c-923a-4f32-9d32-4067a99e76e2 format=QCOW2 isextractable=false passwordEnabled=false isdynamicallyscalable=false ostypeid=528a9a3d-4e13-11ea-8448-1e006e01076c hypervisor=KVM ispublic=false requireshvm=true isfeatured=false isrouting=true details[0].rootDiskController=osdefault directdownload=true
{
  "count": 1,
  "template": [
    {
      "account": "admin",
      "bits": 0,
      "created": "2020-02-13T09:02:45+0000",
      "crossZones": false,
      "details": {
        "rootDiskController": "osdefault"
      },
      "directdownload": true,
      "displaytext": "NewTemplate",
      "domain": "ROOT",
      "domainid": "527d6f21-4e13-11ea-8448-1e006e01076c",
      "format": "QCOW2",
      "hypervisor": "KVM",
      "id": "4cdf991c-4ed8-46f5-be3b-9810c9fce17d",
      "isdynamicallyscalable": false,
      "isextractable": false,
      "isfeatured": false,
      "ispublic": false,
      "isready": true,
      "name": "NewTemplate",
      "ostypeid": "528a9a3d-4e13-11ea-8448-1e006e01076c",
      "ostypename": "Debian GNU/Linux 5.0 (64-bit)",
      "passwordenabled": false,
      "requireshvm": true,
      "size": 328521792,
      "sshkeyenabled": false,
      "status": "Bypassed Secondary Storage",
      "tags": [],
      "templatetype": "ROUTING",
      "zoneid": "94ecae8c-923a-4f32-9d32-4067a99e76e2",
      "zonename": "pr3731-t985-kvm-centos7"
    }
  ]
}

```

For testing databse is updated manually to make newly added template to the SYSTEM type template. This can be done using upgrade script during upgrade,
```
MariaDB [cloud]> update vm_template set type='SYSTEM' where uuid='4cdf991c-4ed8-46f5-be3b-9810c9fce17d';
Query OK, 0 rows affected (0.00 sec)
Rows matched: 1  Changed: 0  Warnings: 0

MariaDB [cloud]> select * from vm_template where uuid='4cdf991c-4ed8-46f5-be3b-9810c9fce17d'\G;
*************************** 1. row ***************************
                  id: 207
         unique_name: 207-2-b5793aa9-d8d1-33bd-b1c5-d144f98a6719
                name: NewTemplate
                uuid: 4cdf991c-4ed8-46f5-be3b-9810c9fce17d
              public: 0
            featured: 0
                type: SYSTEM
                 hvm: 1
                bits: 64
                 url: http://staging.yadav.xyz/systemvmtemplate/4.14/systemvmtemplate-4.14.0-kvm.qcow2.bz2
              format: QCOW2
             created: 2020-02-13 09:02:45
             removed: NULL
          account_id: 2
            checksum: NULL
        display_text: NewTemplate
     enable_password: 0
       enable_sshkey: 0
         guest_os_id: 15
            bootable: 1
         prepopulate: 0
         cross_zones: 0
         extractable: 0
     hypervisor_type: KVM
  source_template_id: NULL
        template_tag: NULL
            sort_key: 0
                size: 328521792
               state: Active
        update_count: 0
             updated: NULL
dynamically_scalable: 0
  parent_template_id: NULL
     direct_download: 1
1 row in set (0.00 sec)
```
Existing SSVM and CPVM are destroyed for testing new template and server can provision new VMs with newly added template.
New VMs get created,
Screenshot from 2020-02-13 14-34-16
Database values for template_id verified in vm_instance,
MariaDB [cloud]> select id,name,vm_template_id from vm_instance where removed IS NULL\G;
```
*************************** 1. row ***************************
            id: 5
          name: VM-17980201-ff14-48c5-ba8a-d90e1be13490
vm_template_id: 4
*************************** 2. row ***************************
            id: 6
          name: r-6-VM
vm_template_id: 203
*************************** 3. row ***************************
            id: 12
          name: s-12-VM
vm_template_id: 207
*************************** 4. row ***************************
            id: 13
          name: v-13-VM
vm_template_id: 207
4 rows in set (0.00 sec)
```

To check direct download template for VR `router.template.*` global config is updated and VR is recreated or new network is created.
```
> list routers filter=id,name
{
  "count": 1,
  "router": [
    {
      "id": "e003704c-c98e-4e84-9a6d-8bfeb45b10cb",
      "name": "r-15-VM"
    }
  ]
}

```
Verified in DB,
```
MariaDB [cloud]> select id,name,vm_template_id from vm_instance where uuid='e003704c-c98e-4e84-9a6d-8bfeb45b10cb'\G;
*************************** 1. row ***************************
            id: 15
          name: r-15-VM
vm_template_id: 207
1 row in set (0.00 sec)
```

### CODE REVIEW

File: core/src/main/java/org/apache/cloudstack/agent/directdownload/HttpDirectDownloadCommand.java [hunk 1]
```diff
@@ -18,14 +18,16 @@
...
 import java.util.Map;
 
+import org.apache.cloudstack.storage.to.PrimaryDataStoreTO;
+
 public class HttpDirectDownloadCommand extends DirectDownloadCommand {
 
-    public HttpDirectDownloadCommand(String url, Long templateId, PrimaryDataStoreTO destPool, String checksum, Map<String, String> headers) {
+    public HttpDirectDownloadCommand(String url, Long templateId, PrimaryDataStoreTO destPool, String checksum, Map<String, String> headers, int connectTimeout, int soTimeout) {
         super(url, templateId, destPool, checksum, headers);
+        setConnectTimeout(connectTimeout);
```
Comment (nvazquez) [inline core/src/main/java/org/apache/cloudstack/agent/directdownload/HttpDirectDownloadCommand.java]: Can these be set on the DrectDownloadCommand constructor?

File: server/src/main/java/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java [hunk 2]
```diff
@@ -320,14 +325,38 @@ private void logUsageEvent(VMTemplateVO template, long poolId) {
...
     private DirectDownloadCommand getDirectDownloadCommandFromProtocol(DownloadProtocol protocol, String url, Long templateId, PrimaryDataStoreTO destPool,
                                                                        String checksum, Map<String, String> httpHeaders) {
+        int connectTimeout = DEFAULT_DIRECT_DOWNLOAD_CONNECT_TIMEOUT;
+        int soTimeout = DEFAULT_DIRECT_DOWNLOAD_SOCKET_TIMEOUT;
+        int connectionRequestTimeout = DEFAULT_DIRECT_DOWNLOAD_CONNECTION_REQUEST_TIMEOUT;
+        if (DownloadProtocol.HTTP.equals(protocol) ||
+                DownloadProtocol.HTTPS.equals(protocol) ||
+                DownloadProtocol.METALINK.equals(protocol)) {
+            try {
+                connectTimeout = Integer.parseInt(configDao.getValue(DirectDownloadConnectTimeout.key()));
```
Comment (nvazquez) [inline server/src/main/java/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: Please do not use this and use the config key value() method

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 3]
```diff
@@ -2247,7 +2247,7 @@ private void updateTemplate(long templateId, String fileExtension, long zoneId,
     }
 
     private void updateDirectDownloadTemplate(long templateId, long zoneId) {
-        TemplateDataStoreVO templateStore = _tmplStoreDao.findByTemplateZone(templateId, zoneId, DataStoreRole.Image);
+        TemplateDataStoreVO templateStore = _tmplStoreDao.findByTemplate(templateId, DataStoreRole.Image);
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: So we are updating every reference in every zone for the template?

**Comment by nvazquez** [comment]
Thanks @shwstppr, left some comments after reviewing

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 3 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @nvazquez as we do not store `store_id` in `cloud.template_store_ref` table for direct download template, we cannot find the entry in the table for a zone while update. Continued with same behaviour from user direct download templates. Do we need to change this for system direct download templates?

File: core/src/main/java/org/apache/cloudstack/agent/directdownload/HttpDirectDownloadCommand.java (hunk 1 above)
Comment (shwstppr) [reply core/src/main/java/org/apache/cloudstack/agent/directdownload/HttpDirectDownloadCommand.java]: Done

File: server/src/main/java/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java (hunk 2 above)
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java]: Done

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Is it ready @shwstppr @Spaceman1984 ?

**Comment by shwstppr** [comment]
No @rhtyd. This is based on another PR so waiting for it to get merge, then rebase.

**Comment by DaanHoogland** [comment]
> No @rhtyd. This is based on another PR so waiting for it to get merge, then rebase.

#3748

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
would be nice if some (other) heavy KVM user could test this; @wido @weizhouapache ?

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Trillian env fails, please check @shwstppr cc @borisstoyanov I'll remove failed envs.

**Comment by shwstppr** [comment]
@rhtyd Failed with the following error while deploying zone,
`Could not find all networks, networks list found: [u'Nested Mgmt Network']`
Don't think an issue with PR changes, investigating more

**Comment by shwstppr** [comment]
Manual pr-trillian job 1025 built fine @rhtyd

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@shwstppr is this still WIP? if no, please remove tag from the title.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@DaanHoogland yes it still needs some fixing related to the listing of direct download system template after removing adding new zones. Working on it

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -1835,6 +1838,9 @@ public Boolean doInTransaction(final TransactionStatus status) {
                     _vlanDao.remove(vlan.getId());
                 }
 
+                // delete template refs for this zone
+                templateZoneDao.deleteByZoneId(zoneId);
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: @DaanHoogland please review if this is the correct place to cleanup template_zone_ref entries for a zone to be removed? Or should this call be made after the successful removal of DataCenterVO in success block below?
cc @rhtyd

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java]: @shwstppr absolutely only on success. I don't think the zone is still usable if deletion of the zoneVO, but it is not right to do an automatic clean-up of items (templates) in a container (zone) if the container is not disposed of.

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
LGTM after internal review.

**Comment by andrijapanicsb** [comment]
Merging based on 4 x approvals and regression test results (and of course manual testing)

**Pull request merged**